Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Cancel.hs

```diff
@@ -157,26 +163,69 @@ cancel req merchant booking mbActiveSearchTry = do
-    let condition = (driverBookingDuration > transporterConfig.driverTimeSpentOnPickupThresholdOnCancel) && (driverDistToPickup < transporterConfig.driverDistanceToPickupThresholdOnCancel)
-    logInfo $ "Verifying condition1 for cancellation: " <> show condition <> " " <> show driverBookingDuration
-    isChargable <-
-      if not condition
-        then do
-          case booking.distanceToPickup of
-            Just acutalDistanceToPickup ->
-              return $ (acutalDistanceToPickup - driverDistToPickup) > transporterConfig.driverDistanceTravelledOnPickupThresholdOnCancel
-            Nothing -> return condition
-        else return condition
-    when isChargable $ do
-      riderId <- booking.riderId & fromMaybeM (RiderDetailsDoNotExist "BOOKING" booking.id.getId)
-      rider <- QRD.findById riderId >>= fromMaybeM (RiderDetailsNotFound riderId.getId)
-      QRD.updateCancellationDues (rider.cancellationDues + transporterConfig.cancellationFee) riderId
+getCancellationCharges :: SRB.Booking -> Maybe SRide.Ride -> Flow (Maybe PriceAPIEntity)
+getCancellationCharges booking mbRide = do
+  case mbRide of
+    Just ride -> do
+      case ride.cancellationFeeIfCancelled of
+        Just cancellationCharges -> return $ Just PriceAPIEntity {amount = cancellationCharges, currency = booking.currency}
+        Nothing -> customerCancellationChargesCalculation booking (Just ride)
+    Nothing -> return Nothing
+
+-- if (driverBookingDuration < freeCancellationTime) || (isDriverOnride - AdvanceBooking)
+-- then no cancellation fee
+-- let distanceCharges = distanceTravelledByDriver * perMetreCancellationCharge
+-- let timeCharges = timeSpentByDriver * perMinuteCancellationCharge
+-- let percentageOfRideFare = (percentageOfRideFareToBeCharged * rideFare)/100
+-- let timeAndDistanceCharges = distanceCharges + timeCharges
+-- MAX (minCharge, MIN(timeAndDistanceCharges, MIN(percentageOfRideFare, maxCharge)))
+
+customerCancellationChargesCalculation :: SRB.Booking -> Maybe SRide.Ride -> Flow (Maybe PriceAPIEntity)
+customerCancellationChargesCalculation booking mbRide = do
+  transporterConfig <- CCT.findByMerchantOpCityId booking.merchantOperatingCityId (Just (TransactionId (Id booking.transactionId))) >>= fromMaybeM (TransporterConfigNotFound booking.merchantOperatingCityId.getId)
+  if not transporterConfig.canAddCancellationFee
+    then return Nothing
+    else do
+      mbCancellationAndNoShowConfigs <- CQCancellationAndNoShowChargeConfigs.findByMerchantOpCityIdAndServiceTierType booking.merchantOperatingCityId booking.vehicleServiceTier
+      case (mbRide, mbCancellationAndNoShowConfigs) of
+        (Just ride, Just cancellationAndNoShowConfigs) -> do
+          now <- getCurrentTime
+          currDistanceToPickup <- getDistanceToPickup ride
+          let initialDistanceToPickup = booking.distanceToPickup
+              timeSpentByDriver = roundToIntegral $ diffUTCTime now ride.createdAt
```

Comment:
ride.createdAt could differ in case of schedule booking?

Reply:
Yes, added a field in ride table called prevRideEndTime to handle this